pkg/genlib/generator_with_template.go

```diff
@@ -0,0 +1,593 @@
+				return field(state, dupes, buf)
+			}
+		}
+
+		parsedTpl, err := t.Funcs(templateFns).Parse(string(tpl))
+		if err != nil {
+			panic(err)
+		}
+		err = parsedTpl.Execute(buf, nil)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	return &GeneratorWithTemplate{emitFuncs: []emitF{f}}, nil
+
+	// t := template.New("generator")
+
+	// templateFns := template.FuncMap{}
+	// objectKeys := make(map[string]struct{})
+	// templateFieldMap := make(map[string][]byte, len(allIndexes))
+
+	// for _, field := range fields {
+	// 	if err := bindField(cfg, field, fieldMap, objectKeys, templateFieldMap); err != nil {
+	// 		panic(err)
+	// 		return nil, err
+	// 	}
+
+	// 	templateFns[field.Name] = func() string {
+	// 		return field.Name
+	// 	}
+	// }
+
+	// parsedTpl, err := t.Funcs(templateFns).Parse(string(tpl))
+	// if err != nil {
+	// 	panic(err)
+	// }
+
+	// f := func(state *GenState, dupes map[string]struct{}, buf *bytes.Buffer) error {
+	// 	err := parsedTpl.Execute(buf, nil)
+	// 	if err != nil {
+	// 		return err
+	// 	}
+
+	// 	return nil
+	// }
+
+	// return &GeneratorWithTemplate{emitFuncs: []emitF{f}}, nil
+	// ============================================================
+
+	// objectKeys := make(map[string]struct{})
+	// var objectKeysFields map[string]Field
+	// if len(tpl) == 0 {
+	// 	tpl, objectKeysFields = generateTemplateFromField(cfg, fields)
+	// 	for k := range objectKeysFields {
+	// 		objectKeys[objectKeysFields[k].Name] = struct{}{}
+	// 	}
+	// }
+
+	// tokenizer := regexp.MustCompile(`([^{]*)({{\.[^}]+}})*`)
+	// allIndexes := tokenizer.FindAllSubmatchIndex(tpl, -1)
+
+	// var fieldPrefixBuffer []byte
+	// var fieldPrefixPreviousN int
+
+	// orderedField := make([]string, 0, len(allIndexes))
+	// templateFieldMap := make(map[string][]byte, len(allIndexes))
+
+	// for _, loc := range allIndexes {
+	// 	var fieldName []byte
+	// 	var fieldPrefix []byte
+
+	// 	if loc[4] > -1 && loc[5] > -1 {
+	// 		fieldName = tpl[loc[4]+3 : loc[5]-2]
+	// 	}
+
+	// 	if loc[2] > -1 && loc[3] > -1 {
+	// 		fieldPrefix = tpl[loc[2]:loc[3]]
+	// 	}
+
+	// 	if len(fieldName) == 0 {
+	// 		if tpl[fieldPrefixPreviousN] == byte(123) {
+	// 			fieldPrefixBuffer = append(fieldPrefixBuffer, byte(123))
+	// 		} else {
+	// 			fieldPrefixBuffer = append(fieldPrefixBuffer, fieldPrefix...)
+	// 		}
+	// 	} else {
+	// 		fieldPrefixBuffer = append(fieldPrefixBuffer, fieldPrefix...)
+	// 		templateFieldMap[string(fieldName)] = fieldPrefixBuffer
+	// 		orderedField = append(orderedField, string(fieldName))
+	// 		fieldPrefixBuffer = nil
+	// 	}
+
+	// 	fieldPrefixPreviousN = loc[2]
+	// }
+
+	// trailingTemplate = fieldPrefixBuffer
+	// // Preprocess the fields, generating appropriate emit functions
+	// fieldMap := make(map[string]emitF)
+	// for _, field := range fields {
+	// 	if err := bindField(cfg, field, fieldMap, objectKeys, templateFieldMap); err != nil {
+	// 		return nil, err
+	// 	}
+	// }
+
+	// for k := range objectKeysFields {
```

Comment:
this is required for when you have `field.*` or similar: [here](https://github.com/elastic/elastic-integration-corpus-generator-tool/pull/37/files#diff-9e27cdd9458e6fafda86103e129642da8680dd2c3cd87947bb5415f5045bd689R47-R85) we transform that to 0 from 5 `field.keyN`, directly in the the `templateFieldMap`. since you'll get rid of it you have to find a way to mimic the same logic somewhere else.

beware that this indeed is only true if we want to generate a default json template given a fields definition. that would be required only if we get rid of the legacy generator.

Reply:
I added support for this, resolving the conversation.